### PR TITLE
AMD SKU support and Refactor of ib tests

### PR DESCRIPTION
**Pull request by rafsalas19**
- Adding mi200/300 Support
  - GPU BW test
  - XGMI test
  - IB tests
  - conf files
- Refactor IB_write tests for more consistency
- Update docs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -109,7 +111,7 @@ The following are Azure custom checks added to the existing NHC suite of tests:
 | check_gpu_ecc | GPU Mem Errors (ECC) |  20000000 | 20000000 | 20000000 | NA | NA |
 | check_gpu_clock_throttling | GPU Throttle codes assertion | not present | not present | not present | NA | NA |
 | check_nccl_allreduce | GPU NVLink bandwidth | 228 GB/s | 228 GB/s | 460 GB/s | NA | NA |
-| check_ib_bw_gdr | IB device (GDR) bandwidth | 175 GB/s | 175 GB/s | 380 GB/s | NA | NA |
+| check_ib_bw_gdr | IB device (GDR) bandwidth | 180 GB/s | 180 GB/s | 380 GB/s | NA | NA |
```
Comment (moes1) [inline README.md:114]: Why did this get changed to 180?

File: conf/nd96asr_mi200_v4.conf [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+###
+### Hardware checks
+###
+ * || check_hw_cpuinfo 2 96 96
+ * || check_hw_physmem 1857801904kB 1857801904kB 3%
+ * || check_hw_swap 0kB 0kB 3%
+ * || check_hw_ib 200 mlx5_ib0:1
+ * || check_hw_ib 200 mlx5_ib1:1
+ * || check_hw_ib 200 mlx5_ib2:1
+ * || check_hw_ib 200 mlx5_ib3:1
```
Comment (moes1) [inline conf/nd96asr_mi200_v4.conf:28]: Is 200 a bit aggressive? That's the limit of the interconnect right?

File: conf/nd96isr_mi300x_v5.conf [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+###
+### Hardware checks
+###
+ *|| check_hw_cpuinfo 2 96 96
+ *|| check_hw_physmem 1915059MB 1915059MB 7%
+ *|| check_hw_swap 0kB 0kB 3%
+ *|| check_hw_ib 400 mlx5_ib0:1
+ *|| check_hw_ib 400 mlx5_ib1:1
+ *|| check_hw_ib 400 mlx5_ib2:1
+ *|| check_hw_ib 400 mlx5_ib3:1
```
Comment (moes1) [inline conf/nd96isr_mi300x_v5.conf:28]: Same question as for mi200.

File: customTests/azure_common.nhc [hunk 4]
```diff
@@ -47,3 +47,59 @@ function remove_clock_boost(){
...
+   return 0
+}
+
+function map_IB_to_numa(){
+   ib_numa=()
+   ib_dev=$(determine_IB_devices)
+
+   IFS=' ' read -ra ib_dev <<< "$ib_dev"
+   for device in "${ib_dev[@]}"; do
+      ib_numa+=($(cat /sys/class/infiniband/$device/device/numa_node))
```
Comment (moes1) [inline customTests/azure_common.nhc]: should there be a check for if this location exists?

File: customTests/azure_gpu_bandwidth.nhc [hunk 5]
```diff
@@ -173,8 +169,187 @@ function check_gpu_bw()
...
+}
+
+function check_rocm_gpu_bw(){
+    # Check GPU BW using ROCM Bandwidth Test 
+    exp_hd_bw=$1
+    exp_dd_bw=$2
+    exp_oam_bw=$3
+
+    if [ ! -f /opt/rocm/bin/rocm-bandwidth-test ]; then
+        die 1 "check_gpu_bw: rocm_bandwidth_test not found. Ensure ROCM is installed and sourced correctly."
```
Comment (moes1) [inline customTests/azure_gpu_bandwidth.nhc]: might want to specifically call out that you may need to install rocm_bandwidth_test if ROCm is installed correctly. Also ROCM -> ROCm.

File: customTests/azure_gpu_bandwidth.nhc [hunk 6]
```diff
@@ -173,8 +169,187 @@ function check_gpu_bw()
...
     return 0
 }
 
+function get_OAM_pairs(){
+    # Get OAM pairs
+    # return array where the index is the GPU id and the value is the OAM pair
+    oam_check=$(rocm-smi --showserial | grep GPU)
+
+    if [[ "$oam_check" == *"Not supported"* ]]; then
+        echo ""
```
Comment (moes1) [inline customTests/azure_gpu_bandwidth.nhc:183]: Need a message here?

File: customTests/azure_gpu_bandwidth.nhc [hunk 7]
```diff
@@ -173,8 +169,187 @@ function check_gpu_bw()
...
+        values="${device_results[$i]}"
+        IFS=$' ' read -ra values <<< "$values"
+        id=${values[0]}
+
+        for ((devid=0; devid<$(( ${#values[*]} -1 )); devid++)) 
+        do
+            val=${values[$(( devid + 1 ))]} #offset first val
+            real_GPUID=$(( devid - CPUNum )) # offset by number of CPUs
+            cur_real_GPUID=$(( id - CPUNum )) # offset by number of CPUs for GPU being evaluated
+            dbg "Rocm Device id $id to Device id $devid: $val GB/s"
```
Comment (moes1) [inline customTests/azure_gpu_bandwidth.nhc]: ROCm

File: README.md [hunk 8]
```diff
@@ -139,6 +141,10 @@ This project has adopted the [Microsoft Open Source Code of Conduct](https://ope
 For more information see the [Code of Conduct FAQ](https://opensource.microsoft.com/codeofconduct/faq/) or
 contact [opencode@microsoft.com](mailto:opencode@microsoft.com) with any additional questions or comments.
 
+### Developer Guide ###
+
+See the [Dev guide](./developer_guide.md) for how if your planning on contributing to this project.
```
Comment (yakovdyadkin) [inline README.md]: Grammar

File: README.md (hunk 1 above)
Comment (rafsalas19) [reply README.md:114]: We weren't boosting clocks before and now we are. With the clocks boosted the results are consistently higher

File: customTests/azure_common.nhc (hunk 4 above)
Comment (rafsalas19) [reply customTests/azure_common.nhc]: Good catch. Will address.

File: customTests/azure_gpu_bandwidth.nhc (hunk 5 above)
Comment (rafsalas19) [reply customTests/azure_gpu_bandwidth.nhc]: Thanks! Will address

File: customTests/azure_gpu_bandwidth.nhc (hunk 6 above)
Comment (rafsalas19) [reply customTests/azure_gpu_bandwidth.nhc:183]: Embarrassing. Yea I will. Thanks!

File: customTests/azure_gpu_bandwidth.nhc (hunk 7 above)
Comment (rafsalas19) [reply customTests/azure_gpu_bandwidth.nhc]: I'll go through all docs and make sure its right

File: README.md (hunk 8 above)
Comment (rafsalas19) [reply README.md]: Thanks! will address

File: customTests/azure_gpu_bandwidth.nhc (hunk 5 above)
Comment (rafsalas19) [reply customTests/azure_gpu_bandwidth.nhc]: I do install it with AZ NHC. Let me specify to check the install.

File: customTests/azure_gpu_bandwidth.nhc (hunk 6 above)
Comment (rafsalas19) [reply customTests/azure_gpu_bandwidth.nhc:183]: Actually, I forgot I meant to do this. Let me add a comment. Essentially if OAM is not supported I just want to echo an empty return value.

File: customTests/azure_common.nhc (hunk 4 above)
Comment (rafsalas19) [reply customTests/azure_common.nhc]: addressed

File: customTests/azure_gpu_bandwidth.nhc (hunk 5 above)
Comment (rafsalas19) [reply customTests/azure_gpu_bandwidth.nhc]: addressed

File: customTests/azure_gpu_bandwidth.nhc (hunk 7 above)
Comment (rafsalas19) [reply customTests/azure_gpu_bandwidth.nhc]: addressed

File: README.md (hunk 8 above)
Comment (rafsalas19) [reply README.md]: addressed

File: customTests/azure_common.nhc (hunk 4 above)
Comment (rafsalas19) [reply customTests/azure_common.nhc]: Addressed

**Review by moes1 — APPROVED** [review]

**Pull request merged**